### PR TITLE
Improve xsbti.Maybe

### PR DESCRIPTION
**Pull request by dwijnand**
* Make Nothing a singleton
* Make Maybe's constructor private
* Optimise equals to use reference equality first
* Optimise Just.equals by having proper, non-anonymous subclasses
* Having non-anonymous subclasses makes them have nicer classnames
* Give Just a value() method

Optimised for reviewing with whitespace changes removed:
https://github.com/sbt/util/pull/55/files?w=1

### CODE REVIEW

File: internal/util-interface/src/main/java/xsbti/Maybe.java [hunk 1]
```diff
@@ -4,46 +4,45 @@
...
+    public String toString() { return "Maybe.just(" + v + ")"; }
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || o instanceof Just<?>) return false;
+      final Just<?> that = (Just<?>) o;
+      return v == null ? that.v == null : v.equals(that.v);
+    }
+  }
+
+  public static final class Nothing extends Maybe<Object> {
```
Comment (Duhemm) [inline internal/util-interface/src/main/java/xsbti/Maybe.java:33]: I think the best way to implement singleton in Java is to use `enum`s?

**Comment by dwijnand** [comment]
Optimised for reviewing with whitespace changes removed: https://github.com/sbt/util/pull/55/files?w=1

File: internal/util-interface/src/main/java/xsbti/Maybe.java [hunk 2]
```diff
@@ -4,46 +4,45 @@
...
+
+    public s value() { return v; }
+
+    public boolean isDefined() { return true; }
+    public s get() { return v; }
+    public int hashCode() { return 17 + (v == null ? 0 : v.hashCode()); }
+    public String toString() { return "Maybe.just(" + v + ")"; }
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || o instanceof Just<?>) return false;
```
Comment (xuwei-k) [inline internal/util-interface/src/main/java/xsbti/Maybe.java]:
``` java
if (o == null || !(o instanceof Just<?>)) return false;
```

**Review by xuwei-k — CHANGES_REQUESTED** [review]

File: internal/util-interface/src/main/java/xsbti/Maybe.java (hunk 2 above)
Comment (dwijnand) [reply internal/util-interface/src/main/java/xsbti/Maybe.java]: Thank you Yoshida-san

File: internal/util-interface/src/main/java/xsbti/Maybe.java (hunk 1 above)
Comment (dwijnand) [reply internal/util-interface/src/main/java/xsbti/Maybe.java:33]: enum's can't extend classes and I don't want to make Maybe an interface

**Comment by dwijnand** [comment]
I'm sure if there's a problem here we can forward-fix or revert.

**Pull request merged**